File: backend/src/services/dashboardService.js
const projectModel = require('../models/projectModel');
const geoModel = require('../models/geoModel');

/**
 * Calculates dynamic dashboard statistics based on authenticated user's role and data scope
 */
async function getDashboardStats(user) {
  const projects = await projectModel.findScoped(user);
  const states = await geoModel.getStates();
  const districts = await geoModel.getDistricts(user.state_id || null);

  // Common calculations
  const totalProjects = projects.length;
  const totalAreaProposed = projects.reduce((sum, p) => sum + parseFloat(p.proposed_area || 0), 0);

  // Calculate acquired area based on status weights
  let totalAreaAcquired = 0;
  projects.forEach(p => {
    const area = parseFloat(p.proposed_area || 0);
    if (p.status === 'POSSESSION_HANDED_OVER') {
      totalAreaAcquired += area;
    } else if (p.status === 'POSSESSION_IN_PROGRESS') {
      totalAreaAcquired += area * 0.75;
    } else if (p.status === 'COMPENSATION_IN_PROGRESS') {
      totalAreaAcquired += area * 0.50;
    } else if (p.status === 'AWARD_IN_PROGRESS') {
      totalAreaAcquired += area * 0.35;
    } else if (p.status === 'NOTIFICATION_IN_PROGRESS') {
      totalAreaAcquired += area * 0.15;
    }
  });

  const acquisitionPercentage = totalAreaProposed > 0 ? ((totalAreaAcquired / totalAreaProposed) * 100).toFixed(1) : '0.0';

  // Dynamic realistic compensation metrics (₹ Cr)
  const compensationAssessed = (totalAreaProposed * 0.45).toFixed(2); // ~45 Lakh/hectare average benchmark
  const compensationPaid = (totalAreaAcquired * 0.42).toFixed(2);

  // Demographic impact metrics
  const affectedFamilies = Math.round(totalAreaProposed * 3.8);
  const displacedFamilies = Math.round(totalAreaProposed * 1.2);
  const rrProgress = totalAreaProposed > 0 ? Math.min(100, Math.round((totalAreaAcquired / totalAreaProposed) * 92)) : 0;

  // Delayed projects
  const delayedProjects = projects.filter(p => p.status === 'DELAYED');

  // Status breakdown
  const statusCounts = {
    PROPOSED: projects.filter(p => p.status === 'PROPOSED').length,
    SURVEY_IN_PROGRESS: projects.filter(p => p.status === 'SURVEY_IN_PROGRESS').length,
    NOTIFICATION_IN_PROGRESS: projects.filter(p => p.status === 'NOTIFICATION_IN_PROGRESS').length,
    AWARD_IN_PROGRESS: projects.filter(p => p.status === 'AWARD_IN_PROGRESS').length,
    COMPENSATION_IN_PROGRESS: projects.filter(p => p.status === 'COMPENSATION_IN_PROGRESS').length,
    POSSESSION_IN_PROGRESS: projects.filter(p => p.status === 'POSSESSION_IN_PROGRESS').length,
    POSSESSION_HANDED_OVER: projects.filter(p => p.status === 'POSSESSION_HANDED_OVER').length,
    DELAYED: delayedProjects.length
  };

  // 1. CENTRAL MINISTRY STATS
  if (user.role === 'CENTRAL_MINISTRY') {
    // State-wise breakdown
    const stateWiseProgress = states.map(st => {
      const stateProjects = projects.filter(p => p.state_id === st.id);
      const stProposed = stateProjects.reduce((sum, p) => sum + parseFloat(p.proposed_area || 0), 0);
      let stAcquired = 0;
      stateProjects.forEach(p => {
        const a = parseFloat(p.proposed_area || 0);
        if (p.status === 'POSSESSION_HANDED_OVER') stAcquired += a;
        else if (p.status === 'POSSESSION_IN_PROGRESS') stAcquired += a * 0.75;
        else if (p.status === 'COMPENSATION_IN_PROGRESS') stAcquired += a * 0.50;
        else if (p.status === 'AWARD_IN_PROGRESS') stAcquired += a * 0.35;
        else if (p.status === 'NOTIFICATION_IN_PROGRESS') stAcquired += a * 0.15;
      });
      const stDelayed = stateProjects.filter(p => p.status === 'DELAYED').length;
      const stPct = stProposed > 0 ? ((stAcquired / stProposed) * 100).toFixed(1) : '0.0';

      return {
        state_id: st.id,
        state_name: st.name,
        state_code: st.code,
        projects_count: stateProjects.length,
        land_proposed: stProposed.toFixed(2),
        land_acquired: stAcquired.toFixed(2),
        acquisition_percentage: stPct,
        delayed_count: stDelayed
      };
    }).filter(s => s.projects_count > 0);

    return {
      role: user.role,
      summary: {
        total_projects: totalProjects,
        total_land_proposed: totalAreaProposed.toFixed(2),
        total_land_acquired: totalAreaAcquired.toFixed(2),
        acquisition_percentage: acquisitionPercentage,
        compensation_assessed_cr: compensationAssessed,
        compensation_paid_cr: compensationPaid,
        affected_families: affectedFamilies,
        displaced_families: displacedFamilies,
        rr_progress_pct: rrProgress,
        delayed_projects: delayedProjects.length,
        average_delay_months: 4.2
      },
      status_counts: statusCounts,
      state_wise_progress: stateWiseProgress,
      delayed_projects_list: delayedProjects,
      recent_activities: getDemoActivities(projects),
      alerts: [
        { id: 1, severity: 'HIGH', message: 'Thane-Borivali Twin Tunnel project requires inter-ministerial forest clearance review.', time: '2 hours ago' },
        { id: 2, severity: 'MEDIUM', message: 'Quarterly compensation audit pending for Lucknow spur segment.', time: '5 hours ago' },
        { id: 3, severity: 'LOW', message: 'National GIS baseline sync scheduled for 02:00 AM IST.', time: '1 day ago' }
      ]
    };
  }

  // 2. STATE GOVERNMENT STATS
  if (user.role === 'STATE_GOVERNMENT') {
    const districtWiseProgress = districts.map(dst => {
      const distProjects = projects.filter(p => p.district_id === dst.id);
      const dProposed = distProjects.reduce((sum, p) => sum + parseFloat(p.proposed_area || 0), 0);
      let dAcquired = 0;
      distProjects.forEach(p => {
        const a = parseFloat(p.proposed_area || 0);
        if (p.status === 'POSSESSION_HANDED_OVER') dAcquired += a;
        else if (p.status === 'POSSESSION_IN_PROGRESS') dAcquired += a * 0.75;
        else if (p.status === 'COMPENSATION_IN_PROGRESS') dAcquired += a * 0.50;
        else if (p.status === 'AWARD_IN_PROGRESS') dAcquired += a * 0.35;
        else if (p.status === 'NOTIFICATION_IN_PROGRESS') dAcquired += a * 0.15;
      });
      return {
        district_id: dst.id,
        district_name: dst.name,
        district_code: dst.code,
        projects_count: distProjects.length,
        land_proposed: dProposed.toFixed(2),
        land_acquired: dAcquired.toFixed(2),
        acquisition_percentage: dProposed > 0 ? ((dAcquired / dProposed) * 100).toFixed(1) : '0.0',
        delayed_count: distProjects.filter(p => p.status === 'DELAYED').length
      };
    }).filter(d => d.projects_count > 0);

    return {
      role: user.role,
      state_name: user.state_name || 'Assigned State',
      summary: {
        state_projects: totalProjects,
        land_proposed: totalAreaProposed.toFixed(2),
        land_acquired: totalAreaAcquired.toFixed(2),
        acquisition_percentage: acquisitionPercentage,
        compensation_assessed_cr: compensationAssessed,
        compensation_paid_cr: compensationPaid,
        affected_families: affectedFamilies,
        displaced_families: displacedFamilies,
        rr_progress_pct: rrProgress,
        delayed_projects: delayedProjects.length
      },
      status_counts: statusCounts,
      district_wise_progress: districtWiseProgress,
      pending_approvals: [
        { id: 'APP-MH-104', project: 'Pune Ring Road Corridor', item: 'Section 19 Declaration Approval', days_pending: 4 },
        { id: 'APP-MH-109', project: 'Pune-Nashik Rail Corridor', item: 'Rehabilitation Grant Sanction', days_pending: 9 }
      ],
      delayed_projects_list: delayedProjects,
      recent_activities: getDemoActivities(projects)
    };
  }

  // 3. DISTRICT AUTHORITY STATS
  if (user.role === 'DISTRICT_AUTHORITY') {
    return {
      role: user.role,
      district_name: user.district_name || 'Assigned District',
      state_name: user.state_name || 'Assigned State',
      summary: {
        district_projects: totalProjects,
        land_proposed: totalAreaProposed.toFixed(2),
        land_acquired: totalAreaAcquired.toFixed(2),
        acquisition_percentage: acquisitionPercentage,
        pending_verification: 14,
        pending_notifications: 3,
        pending_awards: 2,
        compensation_disbursed_pct: 68.4,
        affected_families: affectedFamilies,
        rr_status: 'On Schedule'
      },
      status_counts: statusCounts,
      projects_list: projects,
      field_verification_queue: [
        { id: 'FV-801', parcel_no: 'Survey No. 142/3A', village: 'Haveli', project_name: 'Pune Ring Road', officer: 'Suresh Patil', status: 'PENDING_SURVEY' },
        { id: 'FV-802', parcel_no: 'Gat No. 89/1B', village: 'Mulshi', project_name: 'Pune Ring Road', officer: 'Suresh Patil', status: 'BOUNDARIES_FLAGGED' },
        { id: 'FV-803', parcel_no: 'Survey No. 204/1', village: 'Khed', project_name: 'Pune-Nashik Rail', officer: 'Field Unit 2', status: 'OBJECTION_RECEIVED' }
      ],
      recent_activities: getDemoActivities(projects)
    };
  }

  // 4. PROJECT AGENCY STATS
  if (user.role === 'PROJECT_AGENCY') {
    return {
      role: user.role,
      agency_name: user.agency_name || 'Implementing Agency',
      summary: {
        my_projects: totalProjects,
        land_required: totalAreaProposed.toFixed(2),
        land_acquired: totalAreaAcquired.toFixed(2),
        acquisition_percentage: acquisitionPercentage,
        projects_on_track: totalProjects - delayedProjects.length,
        delayed_projects: delayedProjects.length,
        pending_clearances: 4
      },
      status_counts: statusCounts,
      projects_list: projects,
      milestones: [
        { project_id: 1, name: 'Pune Ring Road', milestone: 'Section 11 Gazetting', progress: 100, status: 'COMPLETED' },
        { project_id: 1, name: 'Pune Ring Road', milestone: 'Joint Measurement Survey (JMS)', progress: 85, status: 'IN_PROGRESS' },
        { project_id: 1, name: 'Pune Ring Road', milestone: 'Final Award & Compensation', progress: 45, status: 'IN_PROGRESS' },
        { project_id: 5, name: 'Ahmedabad-Dholera Exp.', milestone: 'Right of Way Handover', progress: 75, status: 'IN_PROGRESS' }
      ],
      recent_activities: getDemoActivities(projects)
    };
  }

  // 5. FIELD OFFICER STATS
  if (user.role === 'FIELD_OFFICER') {
    return {
      role: user.role,
      district_name: user.district_name || 'Assigned District',
      summary: {
        assigned_projects: totalProjects,
        pending_verification: 6,
        completed_verification: 18,
        assigned_parcels: 24,
        todays_tasks: 3,
        verification_accuracy: '98.5%'
      },
      assigned_projects_list: projects,
      field_tasks: [
        {
          id: 'TSK-101',
          project_name: 'Pune Ring Road Express Corridor',
          village: 'Haveli (Gat No. 142/3A)',
          task_type: 'Ground Boundary Delineation',
          status: 'PENDING',
          gps_coords: '18.5204° N, 73.8567° E',
          due_date: 'Today, 05:00 PM',
          priority: 'HIGH'
        },
        {
          id: 'TSK-102',
          project_name: 'Pune Ring Road Express Corridor',
          village: 'Mulshi (Gat No. 89/1B)',
          task_type: 'Structure & Tree Enumeration',
          status: 'IN_PROGRESS',
          gps_coords: '18.5089° N, 73.5122° E',
          due_date: 'Tomorrow',
          priority: 'MEDIUM'
        },
        {
          id: 'TSK-103',
          project_name: 'Pune-Nashik Rail Corridor',
          village: 'Bhosari (Survey No. 45/2)',
          task_type: 'Land Title Document Verification',
          status: 'PENDING',
          gps_coords: '18.6279° N, 73.8475° E',
          due_date: '28 Aug 2026',
          priority: 'NORMAL'
        }
      ],
      recent_activities: [
        { id: 1, action: 'GPS boundary pinned for Gat No. 140/2', time: '10:30 AM Today' },
        { id: 2, action: 'Site photograph and inspection remark uploaded for Parcel #92', time: 'Yesterday' }
      ]
    };
  }

  return { summary: { totalProjects } };
}

function getDemoActivities(projects) {
  return [
    { id: 1, message: `JMS Survey completed for ${projects[0]?.project_name || 'Project #1'}`, timestamp: '35 mins ago', user: 'Field Division 1' },
    { id: 2, message: 'Section 4 Preliminary Notification gazetted in District Gazette', timestamp: '3 hours ago', user: 'District LAO' },
    { id: 3, message: 'Direct Benefit Transfer (DBT) of ₹4.20 Cr disbursed to 34 land titleholders', timestamp: '1 day ago', user: 'Accounts Officer' },
    { id: 4, message: 'Quarterly review report submitted to Ministry of Road Transport & Highways', timestamp: '2 days ago', user: 'Central PMU' }
  ];
}

module.exports = {
  getDashboardStats
};

File: backend/app/services/dashboard_service.py
from sqlalchemy.orm import Session
from app.models.project import Project
from app.models.user import User, State, District
from app.models.parcel import LandParcel
from app.models.field import FieldTask, FieldVisit, FieldVerification
from app.models.document import Document
from app.models.audit import AuditLog, Alert
from app.services.project_service import get_scoped_projects

def get_dashboard_stats(db: Session, user: User):
    projects = get_scoped_projects(db, user)
    states = db.query(State).all()
    districts_query = db.query(District)
    if user.state_id:
        districts_query = districts_query.filter(District.state_id == user.state_id)
    districts = districts_query.all()

    total_projects = len(projects)
    total_area_proposed = sum(p["proposed_area"] for p in projects)

    total_area_acquired = 0.0
    for p in projects:
        area = p["proposed_area"]
        status = p["status"]
        if status == 'POSSESSION_HANDED_OVER':
            total_area_acquired += area
        elif status == 'POSSESSION_IN_PROGRESS':
            total_area_acquired += area * 0.75
        elif status == 'COMPENSATION_IN_PROGRESS':
            total_area_acquired += area * 0.50
        elif status == 'AWARD_IN_PROGRESS':
            total_area_acquired += area * 0.35
        elif status == 'NOTIFICATION_IN_PROGRESS':
            total_area_acquired += area * 0.15

    acquisition_percentage = round((total_area_acquired / total_area_proposed * 100), 1) if total_area_proposed > 0 else 0.0
    compensation_assessed = round(total_area_proposed * 0.45, 2)
    compensation_paid = round(total_area_acquired * 0.42, 2)
    affected_families = int(round(total_area_proposed * 3.8))
    displaced_families = int(round(total_area_proposed * 1.2))
    rr_progress = min(100, int(round((total_area_acquired / total_area_proposed) * 92))) if total_area_proposed > 0 else 0

    delayed_projects = [p for p in projects if p["status"] == 'DELAYED']

    status_counts = {
        "PROPOSED": sum(1 for p in projects if p["status"] == 'PROPOSED'),
        "SURVEY_IN_PROGRESS": sum(1 for p in projects if p["status"] == 'SURVEY_IN_PROGRESS'),
        "NOTIFICATION_IN_PROGRESS": sum(1 for p in projects if p["status"] == 'NOTIFICATION_IN_PROGRESS'),
        "AWARD_IN_PROGRESS": sum(1 for p in projects if p["status"] == 'AWARD_IN_PROGRESS'),
        "COMPENSATION_IN_PROGRESS": sum(1 for p in projects if p["status"] == 'COMPENSATION_IN_PROGRESS'),
        "POSSESSION_IN_PROGRESS": sum(1 for p in projects if p["status"] == 'POSSESSION_IN_PROGRESS'),
        "POSSESSION_HANDED_OVER": sum(1 for p in projects if p["status"] == 'POSSESSION_HANDED_OVER'),
        "DELAYED": len(delayed_projects)
    }

    recent_activities = _get_real_audit_activities(db, user)

    # 1. CENTRAL MINISTRY
    if user.role == 'CENTRAL_MINISTRY':
        state_wise = []
        for st in states:
            st_projects = [p for p in projects if p["state_id"] == st.id]
            if not st_projects:
                continue
            st_proposed = sum(p["proposed_area"] for p in st_projects)
            st_acquired = sum(
                p["proposed_area"] * (1.0 if p["status"] == 'POSSESSION_HANDED_OVER' else 0.75 if p["status"] == 'POSSESSION_IN_PROGRESS' else 0.50 if p["status"] == 'COMPENSATION_IN_PROGRESS' else 0.35 if p["status"] == 'AWARD_IN_PROGRESS' else 0.15 if p["status"] == 'NOTIFICATION_IN_PROGRESS' else 0)
                for p in st_projects
            )
            st_pct = round((st_acquired / st_proposed * 100), 1) if st_proposed > 0 else 0.0
            state_wise.append({
                "state_id": st.id,
                "state_name": st.name,
                "state_code": st.code,
                "projects_count": len(st_projects),
                "land_proposed": round(st_proposed, 2),
                "land_acquired": round(st_acquired, 2),
                "acquisition_percentage": str(st_pct),
                "delayed_count": sum(1 for p in st_projects if p["status"] == 'DELAYED')
            })

        alerts = _get_real_alerts(db, user)

        return {
            "role": user.role,
            "summary": {
                "total_projects": total_projects,
                "total_land_proposed": round(total_area_proposed, 2),
                "total_land_acquired": round(total_area_acquired, 2),
                "acquisition_percentage": str(acquisition_percentage),
                "compensation_assessed_cr": round(compensation_assessed, 2),
                "compensation_paid_cr": round(compensation_paid, 2),
                "affected_families": affected_families,
                "displaced_families": displaced_families,
                "rr_progress_pct": rr_progress,
                "delayed_projects": len(delayed_projects),
                "average_delay_months": 4.2
            },
            "status_counts": status_counts,
            "state_wise_progress": state_wise,
            "delayed_projects_list": delayed_projects,
            "recent_activities": recent_activities,
            "alerts": alerts
        }

    # 2. STATE GOVERNMENT
    if user.role == 'STATE_GOVERNMENT':
        district_wise = []
        for d in districts:
            d_projects = [p for p in projects if p["district_id"] == d.id]
            d_proposed = sum(p["proposed_area"] for p in d_projects)
            d_acquired = sum(
                p["proposed_area"] * (1.0 if p["status"] == 'POSSESSION_HANDED_OVER' else 0.75 if p["status"] == 'POSSESSION_IN_PROGRESS' else 0.50 if p["status"] == 'COMPENSATION_IN_PROGRESS' else 0.35 if p["status"] == 'AWARD_IN_PROGRESS' else 0.15 if p["status"] == 'NOTIFICATION_IN_PROGRESS' else 0)
                for p in d_projects
            )
            d_pct = round((d_acquired / d_proposed * 100), 1) if d_proposed > 0 else 0.0
            district_wise.append({
                "district_id": d.id,
                "district_name": d.name,
                "district_code": d.code,
                "projects_count": len(d_projects),
                "land_proposed": round(d_proposed, 2),
                "land_acquired": round(d_acquired, 2),
                "acquisition_percentage": str(d_pct),
                "delayed_count": sum(1 for p in d_projects if p["status"] == 'DELAYED')
            })

        alerts = _get_real_alerts(db, user)

        return {
            "role": user.role,
            "state_name": user.state.name if user.state else "Maharashtra",
            "summary": {
                "total_projects": total_projects,
                "total_districts": len(districts),
                "total_land_proposed": round(total_area_proposed, 2),
                "total_land_acquired": round(total_area_acquired, 2),
                "acquisition_percentage": str(acquisition_percentage),
                "compensation_assessed_cr": round(compensation_assessed, 2),
                "compensation_paid_cr": round(compensation_paid, 2),
                "rr_progress_pct": rr_progress,
                "delayed_projects": len(delayed_projects)
            },
            "status_counts": status_counts,
            "district_wise_progress": district_wise,
            "delayed_projects_list": delayed_projects,
            "recent_activities": recent_activities,
            "alerts": alerts
        }

    # 3. DISTRICT AUTHORITY
    if user.role == 'DISTRICT_AUTHORITY':
        # Live queries from PostgreSQL field_tasks, field_verifications, and users
        tasks_query = db.query(FieldTask).join(Project, FieldTask.project_id == Project.id)
        if user.district_id:
            tasks_query = tasks_query.filter(Project.district_id == user.district_id)

        all_tasks = tasks_query.all()
        pending_tasks_count = sum(1 for t in all_tasks if t.status != 'SUBMITTED')
        submitted_tasks_count = sum(1 for t in all_tasks if t.status == 'SUBMITTED')

        # Live Field Verification Queue from PostgreSQL
        verifications_query = db.query(FieldVerification).order_by(FieldVerification.verified_at.desc())
        verifications = verifications_query.all()
        total_verifications = len(verifications)

        queue = []
        for v in verifications:
            t = db.query(FieldTask).filter(FieldTask.id == v.task_id).first()
            p = db.query(LandParcel).filter(LandParcel.id == v.parcel_id).first()
            proj = db.query(Project).filter(Project.id == t.project_id).first() if t else None
            vis = db.query(FieldVisit).filter(FieldVisit.id == v.visit_id).first()
            officer = db.query(User).filter(User.id == vis.field_officer_id).first() if vis else None

            # Attached photos/documents
            photos = db.query(Document).filter(
                Document.related_entity.in_(["FIELD_PHOTO", "FIELD_VISIT"]),
                Document.related_entity_id.in_([v.visit_id, v.task_id, v.parcel_id])
            ).all()

            queue.append({
                "id": f"VER-{v.id}",
                "verification_id": v.id,
                "task_id": v.task_id,
                "ulpin": p.ulpin if p else f"Gat No. {v.parcel_id}",
                "parcel_no": p.ulpin if p else f"Gat No. {v.parcel_id}",
                "survey_number": p.survey_number if p else None,
                "village": p.village if p else "Pune Division",
                "project_name": proj.project_name if proj else "Linear Corridor Project",
                "officer": officer.name if officer else "Field Officer",
                "officer_id": officer.id if officer else None,
                "latitude": float(vis.latitude) if vis and vis.latitude else 18.5204,
                "longitude": float(vis.longitude) if vis and vis.longitude else 73.8567,
                "status": "SUBMITTED",
                "verified_at": v.verified_at.strftime("%I:%M %p, %d %b %Y") if v.verified_at else "Just now",
                "remarks": v.remarks,
                "client_event_id": v.client_event_id,
                "checklist_data": v.checklist_data,
                "photos": [
                    {
                        "id": ph.id,
                        "name": ph.document_name,
                        "url": f"/api/v1/documents/{ph.id}/download",
                        "size": ph.file_size
                    }
                    for ph in photos
                ]
            })

        return {
            "role": user.role,
            "district_name": user.district.name if user.district else "Pune District Collectorate",
            "state_name": user.state.name if user.state else "Maharashtra",
            "summary": {
                "district_projects": total_projects,
                "land_proposed": round(total_area_proposed, 2),
                "land_acquired": round(total_area_acquired, 2),
                "acquisition_percentage": str(acquisition_percentage),
                "pending_verification": pending_tasks_count,
                "submitted_verification": submitted_tasks_count,
                "total_verifications": total_verifications,
                "pending_notifications": 3,
                "pending_awards": 2,
                "compensation_disbursed_pct": 68.4,
                "affected_families": affected_families,
                "rr_status": "On Schedule"
            },
            "status_counts": status_counts,
            "projects_list": projects,
            "field_verification_queue": queue,
            "recent_activities": recent_activities
        }

    # 4. PROJECT AGENCY
    if user.role == 'PROJECT_AGENCY':
        return {
            "role": user.role,
            "agency_name": user.agency.name if user.agency else "Implementing Agency",
            "summary": {
                "my_projects": total_projects,
                "land_required": round(total_area_proposed, 2),
                "land_acquired": round(total_area_acquired, 2),
                "acquisition_percentage": str(acquisition_percentage),
                "projects_on_track": total_projects - len(delayed_projects),
                "delayed_projects": len(delayed_projects),
                "pending_clearances": 0
            },
            "status_counts": status_counts,
            "projects_list": projects,
            "recent_activities": recent_activities
        }

    # 5. FIELD OFFICER
    if user.role == 'FIELD_OFFICER':
        db_tasks = db.query(FieldTask).filter(FieldTask.assigned_to_user_id == user.id).all()
        pending_count = sum(1 for t in db_tasks if t.status != 'SUBMITTED')
        completed_count = sum(1 for t in db_tasks if t.status == 'SUBMITTED')
        total_parcels = db.query(LandParcel).count()

        # Query all documents / photos uploaded by this field officer
        db_docs = db.query(Document).filter(Document.uploaded_by == user.id).order_by(Document.created_at.desc()).all()
        photos_count = sum(1 for d in db_docs if d.file_type and "image" in d.file_type.lower())
        docs_count = sum(1 for d in db_docs if d.file_type and "pdf" in d.file_type.lower())

        # Query all visits initiated by this officer
        db_visits = db.query(FieldVisit).filter(FieldVisit.field_officer_id == user.id).order_by(FieldVisit.visit_start.desc()).all()
        
        # Query all verifications submitted by this officer
        db_verifs = db.query(FieldVerification).join(FieldVisit, FieldVerification.visit_id == FieldVisit.id).filter(FieldVisit.field_officer_id == user.id).order_by(FieldVerification.verified_at.desc()).all()

        live_tasks_list = []
        for t in db_tasks:
            p = db.query(LandParcel).filter(LandParcel.id == t.parcel_id).first()
            proj = db.query(Project).filter(Project.id == t.project_id).first()
            live_tasks_list.append({
                "id": f"TSK-{t.id}",
                "task_id": t.id,
                "project_name": proj.project_name if proj else "Corridor Demarcation",
                "village": f"{p.village} ({p.ulpin})" if p else "Inspection Parcel",
                "ulpin": p.ulpin if p else None,
                "parcel_id": t.parcel_id,
                "survey_number": p.survey_number if p else None,
                "task_type": t.task_type,
                "status": t.status,
                "due_date": t.due_date.strftime("%d %b %Y") if t.due_date else "Today, 05:00 PM",
                "priority": t.priority
            })

        mobile_uploads_list = []
        for d in db_docs:
            is_img = bool(d.file_type and "image" in d.file_type.lower())
            mobile_uploads_list.append({
                "id": d.id,
                "document_id": d.id,
                "file_name": d.document_name,
                "file_type": d.file_type,
                "file_size": d.file_size,
                "upload_type": "PHOTO" if is_img else "PDF",
                "url": f"/api/v1/documents/{d.id}/download",
                "related_entity": d.related_entity,
                "related_entity_id": d.related_entity_id,
                "uploaded_by_name": user.name,
                "created_at": d.created_at.strftime("%d %b %Y, %I:%M %p") if d.created_at else "Recently",
                "status": "UPLOADED"
            })

        visits_list = []
        for v in db_visits:
            t = db.query(FieldTask).filter(FieldTask.id == v.task_id).first()
            p = db.query(LandParcel).filter(LandParcel.id == t.parcel_id).first() if t else None
            has_gps = v.latitude is not None and v.longitude is not None
            visits_list.append({
                "id": v.id,
                "visit_id": v.id,
                "task_id": v.task_id,
                "task_type": t.task_type if t else "Field Inspection",
                "ulpin": p.ulpin if p else None,
                "parcel_number": p.ulpin if p else None,
                "village": p.village if p else None,
                "status": v.status,
                "latitude": float(v.latitude) if has_gps else None,
                "longitude": float(v.longitude) if has_gps else None,
                "accuracy_meters": float(v.accuracy_meters) if v.accuracy_meters else None,
                "gps_display": f"{float(v.latitude):.4f}° N, {float(v.longitude):.4f}° E (±{float(v.accuracy_meters):.1f}m)" if has_gps else "GPS not recorded",
                "visit_start": v.visit_start.strftime("%d %b %Y, %I:%M %p") if v.visit_start else "Recently"
            })

        verifs_list = []
        for vf in db_verifs:
            t = db.query(FieldTask).filter(FieldTask.id == vf.task_id).first()
            p = db.query(LandParcel).filter(LandParcel.id == vf.parcel_id).first()
            verifs_list.append({
                "id": vf.id,
                "verification_id": vf.id,
                "task_id": vf.task_id,
                "visit_id": vf.visit_id,
                "ulpin": p.ulpin if p else None,
                "parcel_number": p.ulpin if p else None,
                "village": p.village if p else None,
                "checklist": vf.checklist_data or {},
                "remarks": vf.remarks,
                "client_event_id": vf.client_event_id,
                "device_id": vf.device_id,
                "verified_at": vf.verified_at.strftime("%d %b %Y, %I:%M %p") if vf.verified_at else "Recently",
                "status": "SUBMITTED"
            })

        latest_gps = visits_list[0]["gps_display"] if (visits_list and visits_list[0]["latitude"]) else "GPS not recorded"

        return {
            "role": user.role,
            "district_name": user.district.name if user.district else "Pune Division",
            "summary": {
                "assigned_projects": len(projects),
                "pending_verification": pending_count,
                "completed_verification": completed_count,
                "assigned_parcels": total_parcels,
                "todays_tasks": len(db_tasks),
                "uploaded_photos_count": photos_count,
                "uploaded_documents_count": docs_count,
                "total_mobile_uploads": len(db_docs),
                "total_field_visits": len(db_visits),
                "total_verifications": len(db_verifs),
                "latest_gps": latest_gps
            },
            "assigned_projects_list": projects,
            "field_tasks": live_tasks_list,
            "mobile_uploads": mobile_uploads_list,
            "field_visits": visits_list,
            "field_verifications": verifs_list,
            "recent_activities": recent_activities
        }

    return {"summary": {"total_projects": total_projects}}

def _get_real_audit_activities(db: Session, user: User):
    query = db.query(AuditLog)
    if user.role == "FIELD_OFFICER":
        query = query.filter(AuditLog.user_id == user.id)
    logs = query.order_by(AuditLog.timestamp.desc()).limit(8).all()

    activities = []
    for log in logs:
        actor = db.query(User).filter(User.id == log.user_id).first() if log.user_id else None
        activities.append({
            "id": log.id,
            "action": log.action.replace("_", " "),
            "message": f"[{log.action}] on {log.entity_type or 'Entity'} #{log.entity_id or ''}",
            "user": actor.name if actor else (log.user_role or "System"),
            "timestamp": log.timestamp.strftime("%I:%M %p, %d %b") if log.timestamp else "Recently"
        })
    return activities

def _get_real_alerts(db: Session, user: User):
    query = db.query(Alert)
    if user.state_id:
        query = query.filter((Alert.state_id == user.state_id) | (Alert.state_id == None))
    if user.district_id:
        query = query.filter((Alert.district_id == user.district_id) | (Alert.district_id == None))

    alerts = query.order_by(Alert.created_at.desc()).limit(5).all()
    return [
        {
            "id": a.id,
            "title": a.title,
            "severity": a.severity,
            "message": a.message,
            "time": a.created_at.strftime("%I:%M %p, %d %b") if a.created_at else "Recently"
        }
        for a in alerts
    ]
